staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader.go

```diff
@@ -160,6 +161,14 @@ func allHeaderValues(h http.Header, headerNames []string) []string {
 	return ret
 }
 
+func unescapeExtraKey(encodedKey string) string {
+	key, err := url.PathUnescape(encodedKey) // Decode %-encoded bytes.
```

Comment:
why does PathUnescape work here?

Reply:
PathEscape is under-aggressive for our needs, but PathUnescape blindly converts %-encoded bytes back to regular bytes: https://golang.org/src/net/url/url.go?s=5075:5118#L173

Reply:
Can you show some examples where `url.PathEscape`/`url.QueryEscape` are lacking?

Reply:
> PathEscape is under-aggressive for our needs

that's unfortunate... what characters does it not escape for us?

Reply:
Looks like it's "@", "=", and ":" are unreserved in path components but forbidden in header keys. I'll add explicit tests for that.
QueryEscape is lacking because we want to maximize compatibility with (i.e. minimize escaping of) existing legal header key strings. We don't want to be over-aggressivley escaping legal strings (which unnecessarily garbles legal keys keys being sent from a new client to an old API server).

I set up a playground to test the interactions between new/old clients/servers with various escaping/unescaping algorithms:
https://play.golang.org/p/eors6oEbRpT

Reply:
Does QueryEscape's aggressive escaping break anything?
I'm very skeptical about having to maintain complex encoding logic for the sake of making headers less garbled/readable.

Reply:
Also, can we offload bulk of escaping to Path/QueryEscape and post-process the output to catch any extra characters we care about?

Reply:
> Does QueryEscape's aggressive escaping break anything?

The `'+' -> ' '` seems problematic

Reply:
> The '+' -> ' ' seems problematic
That would be for Unescape, right? QueryEscape is the other way around, which seems fine?

Reply:
> Does QueryEscape's aggressive escaping break anything?
I'm very skeptical about having to maintain complex encoding logic for the sake of making headers less garbled/readable.

Yes, anyone currently using non-alphanumeric strings and mismatched client/server versions. 

> Also, can we offload bulk of escaping to Path/QueryEscape and post-process the output to catch any extra characters we care about?

Only if it's acceptable to permanently garble keys that are percent-encoded by the caller.
Double-encoding seems more complicated/fragile than simply whitelisting all header chars.

Reply:
>>The '+' -> ' ' seems problematic

>That would be for Unescape, right? QueryEscape is the other way around, which seems fine?

If it's acceptable to additional break old clients sending '+'-containing keys to new API servers, sure.